pylabrobot/thermocycling/opentrons_backend_usb.py

```diff
@@ -0,0 +1,338 @@
+    driver=driver,
+    temperature=temperature,
+    hold_time_seconds=hold_time_seconds,
+    hold_time_minutes=None,
+    ramp_rate=ramp_rate,
+    volume=volume,
+  )
+  await wait_for_block_target(driver)
+
+
+class OpentronsThermocyclerUSBBackend(ThermocyclerBackend):
+  """USB backend for the Opentrons GEN-1/GEN-2 Thermocycler."""
+
+  SUPPORTED_USB_IDS = {
+    (0x04D8, 0xED8C),  # legacy Microchip bridge
+    (0x0483, 0xED8D),  # STMicroelectronics bridge seen in newer units
+  }
+
+  def __init__(self):
+    """Create a new USB backend."""
+    super().__init__()
+    if not USE_OPENTRONS_DRIVER:
+      raise RuntimeError("Opentrons thermocycler driver not available") from _import_error
+
+    self._driver: Optional[AbstractThermocyclerDriver] = None
+    self._current_protocol: Optional[Protocol] = None
+    self._loop: Optional[asyncio.AbstractEventLoop] = None
+
+    self._total_cycle_count: Optional[int] = None
+    self._current_cycle_index: Optional[int] = None
+    self._total_step_count: Optional[int] = None
+    self._current_step_index: Optional[int] = None
+
+  def _clear_cycle_counters(self) -> None:
+    self._total_cycle_count = None
+    self._current_cycle_index = None
+    self._total_step_count = None
+    self._current_step_index = None
+
+  async def _execute_cycle_step(
+    self,
+    temperature: float,
+    hold_time_seconds: float,
+    ramp_rate: Optional[float] = None,
+    volume: Optional[float] = None,
+  ) -> None:
+    """Execute a single thermocycler step (uses shared utility)."""
+    assert self._driver is not None
+    await execute_cycle_step(
+      driver=self._driver,
+      temperature=temperature,
+      hold_time_seconds=hold_time_seconds,
+      ramp_rate=ramp_rate,
+      volume=volume,
+    )
+
+  async def _execute_cycles(
+    self,
+    protocol: Protocol,
+    repetitions: int,
+    volume: Optional[float],
+  ) -> None:
+    """Execute cycles of temperature steps directly from protocol (with cycle tracking)."""
+    assert self._driver is not None
+    self._total_cycle_count = repetitions
+    total_steps = 0
+    for stage in protocol.stages:
+      for _ in range(stage.repeats):
+        for step in stage.steps:
+          total_steps += 1
+    self._total_step_count = total_steps
+
+    step_index = 0
+    for rep in range(repetitions):
+      self._current_cycle_index = rep + 1
+      for stage in protocol.stages:
+        for _ in range(stage.repeats):
+          for step in stage.steps:
+            if len(set(step.temperature)) != 1:
+              raise ValueError(
+                f"Opentrons thermocycler only supports a single unique temperature per step, got {set(step.temperature)}"
+              )
+            temperature = step.temperature[0]
+            hold_time = step.hold_seconds
+            ramp_rate = step.rate if step.rate is not None else None
+            step_index += 1
+            self._current_step_index = step_index
+            await execute_cycle_step(
```

Comment:
instead of two variables `step_index` and `self. _current_step_index`, let's just use `self._current_step_index`

same for `total_steps` just above